apps/communications/dialer/test/marionette/keypad_test.js

```diff
@@ -102,231 +105,270 @@ marionette('Dialer > Keypad', function() {
+
+      var typedNumber = typeLongNumber();
+      var expectedNumberAfterDeletion =
+        typedNumber.substr(0, realDelPosition - 1) +
+        typedNumber.substr(realDelPosition);
+
+      var number = subject.client.findElement(selectors.phoneNumber);
+      var args = [number, delPosition];
+      client.executeScript(function (numberElement, delPosition) {
+        numberElement.selectionStart = numberElement.selectionEnd = delPosition;
+        numberElement.click();
+      }, args);
+
+      var del = subject.client.findElement(selectors.del);
+      actions.tap(del).perform();
+      client.waitFor(function() {
+        return (number.getAttribute('value') == expectedNumberAfterDeletion);
+      });
+      assert.ok(true);
     });
-    assert.ok(true);
-  });
 
-  test('Deleting a couple of digits in the middle of the number', function() {
-    typeNumber();
+    test('Deleting a couple of digits in the middle of the number', function() {
+      typeNumber();
+
+      var number = subject.client.findElement(selectors.phoneNumber);
+      number.scriptWith(function (numberElement) {
+        numberElement.selectionStart = numberElement.selectionEnd = 2;
+        numberElement.click();
+      });
 
-    var number = subject.client.findElement(selectors.phoneNumber);
-    number.scriptWith(function (numberElement) {
-      numberElement.selectionStart = numberElement.selectionEnd = 2;
-      numberElement.click();
+      var del = subject.client.findElement(selectors.del);
+      actions.doubleTap(del).perform();
+
+      assert.equal(number.getAttribute('value'), '3');
     });
 
-    var del = subject.client.findElement(selectors.del);
-    actions.doubleTap(del).perform();
+    test('Clearing the number partially by long pressing the delete key',
+    function() {
+      typeNumber();
 
-    assert.equal(number.getAttribute('value'), '3');
-  });
+      var number = subject.client.findElement(selectors.phoneNumber);
+      number.scriptWith(function (numberElement) {
+        numberElement.selectionStart = numberElement.selectionEnd = 2;
+        numberElement.click();
+      });
 
-  test('Clearing the number partially by long pressing the delete key',
-  function() {
-    typeNumber();
+      var del = subject.client.findElement(selectors.del);
+      actions.longPress(del, 1).perform();
 
-    var number = subject.client.findElement(selectors.phoneNumber);
-    number.scriptWith(function (numberElement) {
-      numberElement.selectionStart = numberElement.selectionEnd = 2;
-      numberElement.click();
+      client.waitFor(function() {
+        return (number.getAttribute('value') === '3');
+      });
+      assert.ok(true, 'cleaned the phone number view');
     });
+});
 
-    var del = subject.client.findElement(selectors.del);
-    actions.longPress(del, 1).perform();
+suite('Add dialed number to a contact', function() {
+    function addTestContact(name, telephone) {
+    contacts.addContact({
```

Comment:
As we're writing an integration test, we can perform injections directly in the database. The test will be faster to run then. Here's a way to do it https://github.com/mozilla-b2g/gaia/pull/33594/files#diff-94d7bf620183da7ba4036797cda4d2beR58